Break out policy computation from endpoint pkg

christarazi opened the pull request:
Review commit-by-commit. It is a long PR, but many of the commits are preparatory.

For convenience, here are the main meaty commits:

> ```
> policy/compute: Add policy computation cell
> 
> This cell is responsible for computing policies (specifically
> SelectorPolicy objects) per identity. In future commits, this cell will
> take over the policy computation logic from the endpoint and the
> endpoint will simply fetch the (latest) computed policy that is stored
> in a statedb table (policy-computations).
> ```

> ```
> endpoint: Plumb policy computer
> 
> Now the policy computation code inside the endpoint regeneration logic
> will simply read from the policy-computations table containing the
> most-recently computed SelectorPolicy object.
> 
> The biggest change to the endpoint logic is the new mechanism to wait
> for the relevant policy computation to appear in the statedb table (with
> the correct revision). For now, this is a hardcoded time-limited retry
> where if it fails to find the relevant entry in the statedb table, then
> the regeneration will fail. The regeneration controller will retry upon
> failure automatically which will retry the fetch of the statedb table.
> 
> Another impact of this commit is that the endpoints that are not
> affected by a policy change will not need to be regenerated, unlike
> before.
> ```

---

Depends on:

* https://github.com/cilium/hive/pull/58
* https://github.com/cilium/statedb/pull/106

Fixes: https://github.com/cilium/cilium/issues/7515

christarazi committed:
policy/compute: Add policy computation cell

This cell is responsible for computing policies (specifically
SelectorPolicy objects) per identity. In future commits, this cell will
take over the policy computation logic from the endpoint and the
endpoint will simply fetch the (latest) computed policy that is stored
in a statedb table (policy-computations).

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon: Add policy compute cell to daemon

Hook up the new cell to the daemon.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpoint: Plumb policy computer

Replace the endpoint's inline policy computation with reads from the
policy-computations statedb table. Endpoints that are not affected by a
policy change no longer need to regenerate.

waitForPolicyComputationResult uses a statedb watch channel with a 1s
timeout to block until the compute cell writes the result. AddHold is
called inside the watch loop immediately after the statedb read so that
the policy cannot be detached between the read and user registration.
On failure the loop waits for the replacement entry via the watch channel.

Moves the old-policy detach from setPolicy into the caller
(ComputeSelectorPolicy), so the compute cell controls when detachment
happens rather than it firing synchronously inside the policy cache.

Adds a skippedPolicyRevision mechanism (mirroring the existing
skippedRegenerationLevel) to ensure duplicate regen triggers propagate
the highest policy revision to the pending regeneration.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
identitymanager: Add getter

This is useful for the policy computation cell which will be added in a
subsequent commit. The policy computation cell will need to fetch
identities from the manager in order to figure out which identities are
relevant for a policy computation, hence this commit.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Trigger policy computation on policy import

This commit triggers policy computation asynchronously which will insert
or update the SelectorPolicy for every identity on the node into the
policy-computations table.

Then after this is kicked off, endpoint regeneration is triggered. In
the subsequent commit, endpoint regeneration will read from the
policy-computations table that was just written to.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Define ComputeSelectorPolicy()

This function will be used to compute policy for a given identity,
similar to GetSelectorPolicy(). Going forward however,
GetSelectorPolicy() will only be used in for unit test purposes.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Expand SelectorPolicy interface with lifecycle methods

Add Detach(), GetRevision(), AddHold(), ReleaseHold(), and MaybeDetach()
to the SelectorPolicy interface. Detach and GetRevision are needed by the
policy-computations statedb table added in the next commit.

The remaining methods implement a hold-based lifecycle for selectorPolicy.
Because policy is now computed per-identity (not per-endpoint), multiple
endpoints sharing an identity read the same selectorPolicy from statedb.
An endpoint must acquire a hold between reading the policy and registering
as a user via DistillPolicy. Without this, a concurrent recomputation
could detach the policy while the endpoint is still adopting it.

MaybeDetach conditionally detaches a superseded policy only when no users
or holds remain. The `superseded` flag ensures removeUser doesn't detach
a policy that is still the current entry in statedb.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy/compute: Add script-based test

Finally, add a simple script-based test that asserts that endpoints'
policies are recomputed correctly in a non-unit test fashion. More
production code paths are execercised in this style of test as compared
to the unit tests that validate the policy code.

This is meant to be the basis for many more tests built on top of what
is added here.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy/cell: Trigger policy computation when needed

This can happen after selector mutation, config change, and from named
ports changes to endpoints or pods, as well as a change observed in CEC.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
identitymanager: Add hive script commands

Similar to the previous commit, we need commands that act on the
IdentityManager.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Add hive commands for repository

We want to be able to do list, add, and debug operations on the policy
repository in script-based test.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpoint/regeneration: Define policy revision to wait for

In a subsequent commit, the policy computation logic inside the endpoint
be replaced by the previously added cell to compute policy. This cell
requires the revision to compute policy for to be passed in, which will
be stored in statedb. The endpoint, when regenerating, will fetch the
policy computation from the statedb table. We need to ensure that the
fetch returns the most up-to-date policy computation which is
represented by this new policy revision field.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
identity/cache: Define script commands for the allocator

Allow hive script-based tests to allocate identities, as well as list
identities that have been allocated.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
ipcache: Define and use MetadataBatchAPI interface

This interface is useful to avoid passing the real ipcache object for
tests, especially for script-based tests initialized using hive.

The policy importer cell will be used in subsequent commits in a
script-based tests so it is necessary to remove the dependence on the
real ipcache.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy/cell: Export the policy importer

In subsequent commits, we need to make the importer accessible to
outside packages in order to utilize the importer in script-based tests
that will need to import policies.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy/cell: Add policy importer script commands

A script-based test will be added that will import policies and trigger
the new policy computation cell / endpoint regeneration path.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
testutils: Define fake certificatemanager

This makes it easier to inject the fake into a script-based test outside
of the policy package since the fake is exported.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpointmanager: Define test cell for script-based tests

The test cell of the endpointmanager is the same as the regular
endpointmanager but without the namespace updater, as we don't need that
in the tests for now. In the future, someone could define an interface
for the namespace updater  and then include a mocked/faked
implementation of it if it's useful.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policymap: Create a fake policymap

This fake policymap which is backed by a standard Go map, instead of a
BPF map, is useful in tests. Especially as subsequent commits will add a
script-based test for the policy / endpoint packages, we need to be able
to test that code without relying on BPF operations.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpointmanager: Add script commands

This is useful to drive script-based tests for the policy / endpoint
package.

Also, revert "endpointmanager: clean up unused EndpointManager interface
method" (2790b19a4a08ca1f019eead8cba7251fa2837585).

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpoint: Add deepcopy-like function for an endpoint

In subsequent commits, a script-based tests will be introduced that
create endpoints. So we will need a mechanism that can create an
endpoint based off a "template" endpoint that is constructed in the test
code, without going through the New*() constructors, for simplicity.

This will be utilized especially in script commands.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy/compute: Add script-based commands

These will help drive the script-based test that will be added in
subsequent commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
endpoint: Remove policy revision check

In the new architecture implemented in the previous commits, policy
updates are computed per-identity and written directly to the
policy-computations table. This replaces the old model where each
endpoint computed its own policy independently. A key difference is the
invocation of UpdatePolicy(): previously, it was called for every
endpoint and included a revision check to detect missed updates. Now,
UpdatePolicy() is only called on endpoints affected by a change. Because
of this, the old revision check would incorrectly fail when policy
updates unrelated to an endpoint occurred, triggering false warnings
about missed updates. The check is therefore removed.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Return old policy from policy cache

When swapping in the new policy object inside the policyCache, return
the old policy. This is useful in subsequent commits when the endpoint
will fetch the new policy and will need access to the old policy in
order to call Detach().

It's worth noting that this is also important for performing the detach
operation at the same time as when the endpoint policy code was before
introducing the policy computation cell that exists outside the endpoint
package. In other words, this commit is in attempt to do the detach
immediately after applying the new policy to the endpoint.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Add convenience function for testing

We need to detect the identity selections from a policy change in the
unit tests to emulate the policy importer cell without using it
directly, due to upcoming changes in later commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

gandro commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -158,6 +160,35 @@ func (idm *IdentityManager) remove(identity *identity.Identity) {
+
+// GetAll returns all identities from the manager. The returned slices contains
+// identities that are pointers to a live objects; do not modify!
+func (idm *IdentityManager) GetAll() []*identity.Identity {
```
Have you considered returning an iterator here? It would of course keep the readlock during the iteration, which may or may not be desired.

christarazi replied:
The readlock would need to be taken before the iteration can start, otherwise each time `yield` would be called then the lock will be taken and then released over and over again, right?

gandro replied:
Very true, sounds like a footgun. Thinking more about this it seems that returning a snapshot here is the proper thing to do.

gandro commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -158,6 +160,35 @@ func (idm *IdentityManager) remove(identity *identity.Identity) {
 
+// Get returns the full identity based on the numeric identity. The returned
+// identity is a pointer to a live object; do not modify!
+func (idm *IdentityManager) Get(id *identity.NumericIdentity) *identity.Identity {
```
I guess this exists for symmetry with `Add/Remove` but it seems entirely unused

christarazi replied:
It is used in `(*IdentityPolicyComputer).UpdatePolicy()`

gandro commented on pkg/policy/distillery.go:
```diff
@@ -202,10 +204,6 @@ func (cip *cachedSelectorPolicy) getPolicy() *selectorPolicy {
-	oldPolicy := cip.policy.Swap(policy)
-	if oldPolicy != nil {
-		// Release the references the previous policy holds on the selector cache.
-		oldPolicy.detach(false, endpointID)
```
nit: The doc comment still says that it calls detach on the old policy, which is no longer true

gandro commented on pkg/policy/distillery.go:
```diff
@@ -92,8 +94,8 @@ func (cache *policyCache) delete(identity *identityPkg.Identity) bool {
 // Must be called with repo.Mutex held for reading.
-func (cache *policyCache) updateSelectorPolicy(identity *identityPkg.Identity, endpointID uint64) (*selectorPolicy, bool, error) {
-	cip := cache.lookupOrCreate(identity)
+func (cache *policyCache) updateSelectorPolicy(identity *identityPkg.Identity, endpointID uint64) (*selectorPolicy, *selectorPolicy, bool, error) {
```
nit: consider naming the return arguments here, to clearly indicate which one is the new and the old selectorPolicy without asking the reader to study the function body

gandro commented on pkg/policy/distillery.go:
```diff
@@ -38,15 +38,17 @@ func newPolicyCache(repo *Repository, idmgr identitymanager.IDManager) *policyCa
 // lookupOrCreate adds the specified Identity to the policy cache, with a reference
 // from the specified Endpoint, then returns the threadsafe copy of the policy.
-func (cache *policyCache) lookupOrCreate(identity *identityPkg.Identity) *cachedSelectorPolicy {
+func (cache *policyCache) lookupOrCreate(identity *identityPkg.Identity) (*cachedSelectorPolicy, bool) {
```
nit: consider named return values here as well

gandro commented on pkg/policy/repository.go:
```diff
@@ -37,14 +37,19 @@ type PolicyRepository interface {
 
+	// ComputeSelectorPolicy computes the SelectorPolicy for a given identity,
+	// if it needs recomputing.
+	ComputeSelectorPolicy(id *identity.Identity, skipRevision uint64) (SelectorPolicy, uint64, SelectorPolicy, bool, error)
```
nit: Please name the return arguments here, as it is not clear which one is old and new selector policy

gandro commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+type PolicyRecomputer interface {
+	RecomputeIdentityPolicy(identity *identity.Identity, toRev uint64) (<-chan struct{}, error)
+	RecomputeIdentityPolicyForAllIdentities(toRev uint64) (*statedb.WatchSet, error)
+	UpdatePolicy(idsToRegen *set.Set[identity.NumericIdentity], fromRev, toRev uint64)
```
Why is this taking a pointer to a set? Sets are already passed by reference

gandro commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+
+	ch := make(chan struct{}, 1)
+	go func() {
+		r.mu.Lock()
```
Note that taking a `WriteTxn` also effectively is the same as a mutex, as in `WriteTxn` blocks if there is a concurrent write transaction in progress. So this `r.mu.Lock()` logic that seems to protect against concurrent recomputation seems redundant (unless I'm missing something).

jrajahalme replied:
Not sure of the context here, but a mutex can be needed to block concurrent readers, if there are other fields than the db table that need synchronization, as there is no way to take a read lock on the statedb managed locks. But it looks like the mutex is superfluous here.

gandro commented on pkg/policy/compute/table.go:
```diff
@@ -0,0 +1,63 @@
+
+var (
+	PolicyComputationNameIndex = statedb.Index[Result, identity.NumericIdentity]{
+		Name: "name",
```
I think "numeric-identity" would be better name for this index

gandro commented on pkg/policy/cell/policy_importer.go:
```diff
@@ -385,6 +390,8 @@ func (i *policyImporter) processUpdates(ctx context.Context, updates []*policyty
 	// Unaffected endpoints can merely have their policy revision set.
 	i.log.Info("Policy repository updates complete, triggering endpoint updates",
 		logfields.PolicyRevision, endRevision)
+	fmt.Printf("chris debug: policyImporter calling UpdatePolicy: %v\n", idsToRegen)
```
leftover debug printf

gandro commented on pkg/endpoint/policy.go:
```diff
@@ -194,7 +196,7 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
 
 	// Copy out some values we care about, then unlock
-	forcePolicyCompute := e.forcePolicyCompute
+	// forcePolicyCompute := e.forcePolicyCompute
```
This should probably be removed?

gandro commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+	return ch, nil
+}
+
+// RecomputeIdentityPolicy recomputes the policy for a specific identity.
```
nit: doc comment is wrong

gandro commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -223,3 +230,95 @@ type IdentitiesModel []*models.IdentityEndpoints
+	return map[string]script.Cmd{
+		"idm/list": script.Command(
+			script.CmdUsage{
+				Summary: "List all policies in the policy repository",
```
nit: Wrong summary

gandro commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -223,3 +230,95 @@ type IdentitiesModel []*models.IdentityEndpoints
+		),
+		"idm/add": script.Command(
+			script.CmdUsage{
+				Summary: "List all policies in the policy repository",
```
nit: Wrong summary

gandro commented on pkg/endpointmanager/script_cmds.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+func ScriptCmds(epm EndpointManager, template *endpoint.Endpoint) map[string]script.Cmd {
```
Have you considered using `models.EndpointChangeRequest` as the input here?

christarazi replied:
Yes, I think the issue is that there are dependencies that are created within the scope of the script test and therefore would need to all be passed into this `ScriptCmds` function which would make the function prototype look like the `createEndpoint()`. The `EndpointChangeRequest` is also not sufficient either because it holds more metadata about the endpoint and what's important for us here is the dependencies, not the metadata.

That's why I've left it as the `Endpoint` object itself so all the relevant fields can be initialized.

gandro commented on pkg/policy/cell/policy_importer.go:
```diff
@@ -427,3 +434,108 @@ func truncate[T any](xs iter.Seq[T], maxLen int) iter.Seq[T] {
 }
+
+func PolicyImporterScriptCmds(p *Importer) map[string]script.Cmd {
+	f := func(s *script.State, args []string, del bool) (api.Rules, ipcachetypes.ResourceID, error) {
```
nit: consider giving `f` a more descriptive name

gandro commented on pkg/policy/test/testdata/test.txtar:
```diff
@@ -0,0 +1,88 @@
+
+# Perform assertion that all endpoints with a non-reserved identity (not host /
+# health) are still at policy revision 3.
+db/show --columns Identity,Revision policy-computations
```
Why not `db/cmp` here?

gandro reviewed (commented):
Awesome work! Unfortunately I'm not familiar enough with the endpoint/policy interaction lifecycle to review that aspect in depth. So I only have mostly nits for you

pippolo84 commented on pkg/policy/resolve.go:
```diff
@@ -161,6 +161,9 @@ type SelectorPolicy interface {
 	// DistillPolicy returns the policy in terms of connectivity to peer
 	// Identities.
 	DistillPolicy(logger *slog.Logger, owner PolicyOwner, redirects map[string]uint16) *EndpointPolicy
+
+	Detach()
+	GetRevision() uint64
```
Nit: might be good to add inline doc here since these methods are exported

pippolo84 commented on pkg/policy/compute/cell.go:
```diff
@@ -0,0 +1,79 @@
+		logger:    params.Logger,
+	}
+	params.IDManager.Subscribe(obj)
+
+	params.Lifecycle.Append(cell.Hook{
+		OnStart: func(cell.HookContext) error {
+			params.Logger.Info("Identity policy recomputer started")
+			return nil
+		},
+		OnStop: func(cell.HookContext) error {
+			params.Logger.Info("Identity policy recomputer stopped")
+			return nil
+		},
```
[NON-BLOCKING] These logs are actually redundant, since hive already emits logs for each start/stop hook it runs:

```
time=2025-10-27T18:28:33.099357634Z level=debug source=/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:126 msg="Executing start hook" function="compute.NewIdentityPolicyComputer.func1 (.../policy/compute/cell.go:45) (agent.controlplane.identity-policy-computer)"
time=2025-10-27T18:28:33.099373174Z level=info source=/go/src/github.com/cilium/cilium/pkg/policy/compute/cell.go:46 msg="Identity policy recomputer started" module=agent.controlplane.identity-policy-computer
time=2025-10-27T18:28:33.099389775Z level=debug source=/go/src/github.com/cilium/cilium/vendor/github.com/cilium/hive/cell/lifecycle.go:136 msg="Start hook executed" function="compute.NewIdentityPolicyComputer.func1 (.../policy/compute/cell.go:45) (agent.controlplane.identity-policy-computer)" duration=18.025µs
```

However, I see that the messages might be simpler to read and find. Moreover, it is just an additional log line for each startup and stop events, so feel free to keep them if you think these will increase the usability of the log.

pippolo84 commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+			wtxn.Abort()
+		}
+
+	out:
```
Nit: I suggest the following version since it is a bit more idiomatic (we keep the abort/unlock/close closer to the related initiating operations, plus we use early return that should be easier to read):

```diff
 	ch := make(chan struct{}, 1)
 	go func() {
+		defer close(ch)
+
 		r.mu.Lock()
+		defer r.mu.Unlock()
 
 		wtxn := r.db.WriteTxn(r.tbl)
+		defer wtxn.Abort()
+
 		res := Result{Identity: identity.ID}
 
 		var err error
@@ -59,20 +64,17 @@ func (r *IdentityPolicyComputer) RecomputeIdentityPolicy(identity *identity.Iden
 		// If so, skip calculation.
 		obj, _, found := r.tbl.Get(wtxn, PolicyComputationByIdentity(identity.ID))
 		if found && obj.Revision >= toRev {
-			goto out
+			wtxn.Commit()
+			return
 		}
 
 		res.NewPolicy, res.Revision, res.OldPolicy, res.NeedsRelease, res.Err = r.repo.ComputeSelectorPolicy(identity, toRev)
 		_, _, err = r.tbl.Insert(wtxn, res)
 		if err != nil {
 			r.logger.Error("Failed to write into statedb policy computation table", logfields.Error, err)
-			wtxn.Abort()
+			return
 		}
-
-	out:
 		wtxn.Commit()
-		r.mu.Unlock()
-		close(ch)
 	}()
```

pippolo84 commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+		logfields.PolicyRevision, toRev,
+	)
+
+	ch := make(chan struct{}, 1)
```
Shouldn't an unbuffered channel be sufficient in this case?

christarazi replied:
I thought that an unbuffered channel would block the sender until the receiver receives from the sender? In this case, I want to make sure that the caller is not blocked unless the caller explicitly wants to wait on the channel. Or am I wrong and it's actually the same behavior here?

pippolo84 replied:
Regarding the producer, `close` won't block even if the channel is unbuffered, it basically records that no more values will be sent on the channel (see the [spec](https://go.dev/ref/spec#Close)).
Regarding the consumer, it can choose to discard the channel to avoid being blocked or to wait on the channel until it's closed, no matter if it is buffered or not.
So, yes, with respect to closing and receiving it's the same behavior :+1:

jrajahalme replied:
I think sender/caller would get blocked until the receiver receives from an unbuffered channel. For a buffered channel that happens only if the channel is full.

pippolo84 replied:
> I think sender/caller would get blocked until the receiver receives from an unbuffered channel. For a buffered channel that happens only if the channel is full.

That is correct, but here we are just using `close` to signal the caller and `close` won't block the producer in any way, even if the underlying channel is unbuffered.

pippolo84 reviewed (commented):
Awesome! 💯 

Made a quick pass on the first ~10 commits focusing on aspects related to my codeowner plus a couple of nits.
Will do a second round ASAP.

pippolo84 commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+	defer r.mu.Unlock()
+
+	wtxn := r.db.WriteTxn(r.tbl)
+	defer wtxn.Commit()
```
Nit: the pattern

```go
	wtxn := r.db.WriteTxn(r.tbl)
	defer wtxn.Abort()

	if <error> {
		return
	}

	wtxn.Commit()
```

might be better here too

pippolo84 commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+	"github.com/cilium/cilium/pkg/policy"
+)
+
+type PolicyRecomputer interface {
```
I see that an interface might be useful for testing, but OTOH it makes the code harder to read adding an extra layer of indirection. Given that we already have `InstantiateCellForTesting` and we don't actually mock the computer in any tests, I suggest to just Provide the concrete type `*PolicyIdentityComputer`

pippolo84 commented on pkg/endpoint/policy.go:
```diff
@@ -209,30 +211,29 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
+	pcr, err = e.waitForPolicyComputationResult(datapathRegenCtxt, securityIdentity, result)
+	if err != nil {
+		return err
+	} else if pcr != nil {
```
Nit: no need for the `else` here since we `return` in the `if` before

christarazi replied:
It's possible that the policy computation result itself contains an error and there is not an error returned from the wait function.

jrajahalme replied:
Looks like `else` is still redundant? I.e., this should be the same:
```
	pcr, err = e.waitForPolicyComputationResult(datapathRegenCtxt, securityIdentity, result)
	if err != nil {
		return fmt.Errorf("failed waiting for policy computation result: %w", err)
	}
    if pcr != nil {
```

pippolo84 commented on pkg/endpoint/policy.go:
```diff
@@ -267,12 +268,66 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
+	for {
+		if retries >= 3 {
+			datapathRegenCtxt.policyResult = result
+			return nil, errors.New("failed to fetch computed policy")
```
Should we add the triggering security identity in the error message to ease debugging?

christarazi replied:
The caller will log the error and it will log the identity already, so no need to augment the error msg.

pippolo84 commented on pkg/endpoint/policy.go:
```diff
@@ -267,12 +268,66 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
+		case found && computeResult.Revision >= datapathRegenCtxt.policyRevisionToWaitFor:
+			break done
+		case found && computeResult.Revision < datapathRegenCtxt.policyRevisionToWaitFor:
+			retries++
```
Do we really want to increase retry in case the revision is older than the expected one? AFAIU, in this case we expect to hit the `<-watch` case in the select below, but I'm afraid that if we hit the first case (unlikely, but it might happen) we end up increasing the retries counter twice for a single attempt.

christarazi replied:
Good catch, fixed.

pippolo84 reviewed (commented):
Some other nits and comments here and there

jrajahalme commented on pkg/identity/identitymanager/manager.go:
```diff
@@ -158,6 +160,35 @@ func (idm *IdentityManager) remove(identity *identity.Identity) {
+	return idd.identity
+}
+
+// GetAll returns all identities from the manager. The returned slices contains
```
nit: "slices" -> "slice"

jrajahalme commented on pkg/policy/rules.go:
```diff
@@ -93,6 +95,16 @@ func (rules ruleSlice) AsPolicyEntries() types.PolicyEntries {
 	return policyRules
 }
 
+func (rules ruleSlice) AllIdentitySelections() set.Set[identity.NumericIdentity] {
```
Wonder if this could be an iterator instead?

jrajahalme commented on pkg/policy/compute/compute.go:
```diff
@@ -0,0 +1,137 @@
+
+	r.logger.Info("Recomputing policy for all identities")
+	for _, id := range r.idmanager.GetAll() {
+		if ch, err := r.RecomputeIdentityPolicy(id, toRev); err != nil {
```
What happens to identities added to the identity manager right after `GetAll()` returns? It looks like `LocalEndpointIdentityAdded` would take care of that?

jrajahalme commented on pkg/endpoint/policy.go:
```diff
@@ -267,12 +268,66 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
 	result.endpointPolicy = selectorPolicy.DistillPolicy(e.getLogger(), e, desiredRedirects)
+	if pcr.NeedsRelease {
+		oldPolicy.Detach()
+	}
```
It looks to me now that the old policy should be detached only after the last endpoint using it starts using the new policy, not when the first endpoint starts using the new policy. Additionally, IMO this logic (when to detach the old policy) should not be exposed to the endpoint package.

The difference with detach on first vs. last is that with the current behavior all endpoints using the same selector policy will stop receiving incremental updates when the first endpoint starts using the new selector policy. On a busy system this could lead to spurious policy drops on the other endpoints due to `toFQDN` related (incremental) policy updates not happening until each of the endpoints starts using the new selector policy.

christarazi replied:
I pushed a WIP commit as the last commit that implements this. Could you take a look?

jrajahalme commented on pkg/endpoint/policy.go:
```diff
@@ -267,12 +268,66 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics, datapathRegen
 	return nil
 }
 
+func (e *Endpoint) waitForPolicyComputationResult(
```
It would be useful to have a guarantee that whenever a new or current selector policy is returned (via `Result`), the caller could rely on that policy not being detached while this endpoint is preparing to use it.

This ties in with the other comment on not detaching on the policy change, but on the last user.

Even the "last user" detachment policy has a potential race: What it the last *current* user stops using the policy, while a new endpoint has just received the policy (via this function) and is preparing to start using it? Currently we let the policy attach fail silently, leading to the new endpoint using a detached selector policy and not getting any incremental updates until the endpoint is regenerated again. The minimum we should do to fix this is to return an error up from the `EndpointPolicy.attach` that tries to attach to a detached selector policy, and have the endpoint retry the policy computation immediately.

christarazi replied:
Hm, I don't see a `EndpointPolicy.attach`. Could you point me to where you are thinking this should be? Also, how do we know if the selectorpolicy has already been detached? From the `users` field being empty / length 0?

jrajahalme replied:
`p.insertUser()` happens in `selectorPolicy) DistillPolicy` at `pkg/policy/resolve.go`. The problem is that selectorPolicy can currently be detached before `p.insertUser()` is reached.

jrajahalme replied:
We may need to add a separate `holdCount` to count references to a selectorPolicy so that we can increase the count as soon as a selectorPolicy is computed or reused for an endpoint, since the `EndpointPolicy` does not exist yet. This count should be dropped as soon as the EndpointPolicy is added as the user, and then the detach on the user removal should be skipped if the count is non-zero. Detach could then be done from two places: removal of the last user if `holdCount` is zero and count release function if the count reaches zero and the user count is also zero, both done under a lock?

jrajahalme commented on pkg/policy/repository.go:
```diff
@@ -603,3 +611,88 @@ func (p *Repository) GetPolicySnapshot() map[identity.NumericIdentity]SelectorPo
+		),
+		"policyrepo/selectorcache": script.Command(
+			script.CmdUsage{
+				Summary: "List all policies in the policy repository",
```
Summary is a copy from `policyrepo/list`, needs to be changed to describe `policyrepo/selectorcache`.

jrajahalme reviewed (commented):
This is a really nice improvement into to the right direction. I have some nits, but also a concern about endpoints getting successfully regenerated and using a detached selector policy, or the selector policy getting detached while they are still using it. To be clear, this is an existing concern that has not been introduced in this PR, but it would be nice if this aspect would be fixed here.

christarazi committed:
policy: Define JSON marshaling function for RangePortMap

This is needed when printing the policy-computations table.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
identitymanager: Move subscriber callbacks from critical section

This commit moves the identitymanager's subscriber callbacks out of the
critical section of the mutex, in case the subscribers invoke the
identitymanager themselves and take the read lock. This will be useful
in subsequent commits.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
policy: Make policy cache events an observable stream

This commits makes the policyCache object an observable stream whenever
the cache is modified with policy changes. This is useful in subsequent
commits when the policy computer observes this stream in order to
compute or delete a policy computation based on the policyCache event.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

jrajahalme commented on pkg/policy/resolve.go:
```diff
@@ -297,7 +297,9 @@ func (p *selectorPolicy) insertUser(user *EndpointPolicy) {
-	p.L4Policy.removeUser(user)
+	if p.L4Policy.removeUser(user) {
+		p.detach(true, 0)
+	}
```
This could be racy, `removeUser()` can return `true` and a new user can come in before we reach `p.detach()` here.

jrajahalme replied:
`p.detach()` is just `p.L4Policy.detach()` so maybe `removeUser() could detach while holding the lock:
```
func (l4 *L4Policy) removeUser(user *EndpointPolicy) {
	l4.mutex.Lock()
	if l4.users != nil {
		delete(l4.users, user)
        if len(l4.users) == 0 { // last user removed
            // these lines are from L4Policy.detach()
            l4.Ingress.Detach(selectorCache)
            l4.Egress.Detach(selectorCache)
            l4.users = nil
            // 'users' is empty so the rest of L4Policy.detach() implementation is not needed here
        }
	}
	l4.mutex.Unlock()
}
```

jrajahalme replied:
This still needs to be enhanced so that we can detach also when holdCount gets to zero, maybe factor the innards to a separate maybeDetachLocked() function:
```
func (l4 *L4Policy) maybeDetachLocked() {
    if l4.holdCount == 0 && (l4.users == nil || len(l4.users) == 0) {
            // these lines are from L4Policy.detach()
            l4.Ingress.Detach(selectorCache)
            l4.Egress.Detach(selectorCache)
            l4.users = nil
            // 'users' is empty so the rest of L4Policy.detach() implementation is not needed here
    }
}

func (l4 *L4Policy) removeUser(user *EndpointPolicy) {
	l4.mutex.Lock()
	if l4.users != nil {
		delete(l4.users, user)
	}
    l4.maybeDetachLocked()
	l4.mutex.Unlock()
}

func (l4 *L4Policy) releaseHold() {
	l4.mutex.Lock()
    if l4.holdCount > 0 {
        l4.holdCount--
    }
    l4.maybeDetachLocked()
    l4.mutex.Unlock()
}
```

christarazi replied:
Implemented maybeDetachLocked() as suggested, with one change to the condition: using l4.users != nil && len(l4.users) == 0 instead of l4.users == nil || len(l4.users) == 0.

The reason is that insertUser is called from within DistillPolicy, which holds the SelectorCache read lock. If the policy is already detached (`users == nil`) and maybeDetachLocked is called inside insertUser, it would call Ingress.Detach, which acquires the write lock while the read lock is held, so it would deadlock.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -1606,16 +1606,18 @@ func (l4 *L4Policy) insertUser(user *EndpointPolicy) {
+		last = len(l4.users) == 0
 	}
 	l4.mutex.Unlock()
+	return last
```
For this to not be race the detach may need to be done while holding the lock, see comment below.
